### PR TITLE
Added chop solver to logreg_l2 benchmark

### PR DESCRIPTION
**Pull request by GeoffNN**

### CODE REVIEW

**Comment by GeoffNN** [comment]
Adding a stochastic method as well

File: solvers/chop.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import warnings
+from benchopt import BaseSolver, safe_import_context
+
+with safe_import_context() as import_ctx:
+    import torch
+    import chop
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
```
Comment (tomMoral) [inline solvers/chop.py:9]: This call should be put in `set_objective`

As it is now, it would prevent `benchopt` from importing the class in an emv where pytorch is not installed.
 This is for instance the case when collecting the requirements to install them.

Comment (tomMoral) [reply solvers/chop.py:9]: actually, device is not used?

File: solvers/chop.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+class Solver(BaseSolver):
+    name = 'chop'
+
+    install_cmd = 'conda'
+    requirements = ['pip:https://github.com/openopt/chop/archive/master.zip']
+
+    parameters = {'solver': ['pgd'],
+                  'line_search': [False, True]}
+
+    def skip(self, X, y, lmbd):
```
Comment (tomMoral) [inline solvers/chop.py:20]: By default, you don't need to implement the skip method except if your solver is not compatible with some of the benchmark objective. So if you don't need to skip something, you can remove this.

**Review by tomMoral — COMMENTED** [review]
thanks for the PR @GeoffNN !! this looks very cool.

a few comments on the side. could you also show us the curve you obtain with this solver ?

**Comment by tomMoral** [comment]
Oh and for the tests, it seems that our CI is broken.. I will take a look to fix it asap.

in the mean time, could you run `benchopt test .` in this benchmark to check the solver was passing the tests?

**Comment by agramfort** [comment]
I took the liberty to push here :)

>

File: solvers/chop.py (hunk 1 above)
Comment (GeoffNN) [reply solvers/chop.py]: Now it is!

File: solvers/chop.py (hunk 2 above)
Comment (GeoffNN) [reply solvers/chop.py:28]: Thanks -- now there's a bunch of combinations to skip, depending on the setting: full gradient vs stochastic mini batch

File: solvers/chop.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        x0 = torch.zeros(n_features, dtype=X.dtype, device=X.device)
+        if n_iter == 0:
+            self.beta = x0
+            return
+
+        criterion = torch.nn.BCEWithLogitsLoss()
+
+        warnings.filterwarnings('ignore', category=RuntimeWarning)
+
+        if self.stochastic:
```
Comment (tomMoral) [inline solvers/chop.py:84]: Could you move this two part of the code in separate functions to improve readability?

File: solvers/chop.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        return False, None
+
+    def set_objective(self, X, y, lmbd):
+        self.lmbd = lmbd
+
+        if self.device == 'cpu':
+            os.environ["CUDA_VISIBLE_DEVICES"] = ""
+        elif 'CUDA_VISIBLE_DEVICES' in os.environ:
+            del os.environ["CUDA_VISIBLE_DEVICES"]
```
Comment (tomMoral) [inline solvers/chop.py:64]: Not a big fan of having solver that have an impact on the global state of the interpreter.
Is this really required?

**Review by tomMoral — COMMENTED** [review]
A few more comments.

File: solvers/chop.py (hunk 4 above)
Comment (GeoffNN) [reply solvers/chop.py:69]: @agramfort added this part -- would `device = torch.device(self.device)` be enough?

**Comment by GeoffNN** [comment]
The lightning install test is failing for some reason -- other than that everything passes on my machine (equipped w/ GPU).
The resulting plot is not readable: there's too many labels in the legend box, and the plot doesn't resize.

[plot_benchopt.pdf](https://github.com/benchopt/benchmark_logreg_l2/files/6122347/plot_benchopt.pdf)

**Comment by agramfort** [comment]
yes the plots need to be improved eg putting the legend in the outside

can you try to publish the benchmark result:

See benchopt publish command. REsults will appear in

https://benchopt.github.io/results/benchmark_logreg_l2.html

**Comment by GeoffNN** [comment]
Uploaded!

**Comment by agramfort** [comment]
here is how it looks

https://benchopt.github.io/results/benchmark_logreg_l2_benchopt_run_2021-03-11_14h48m57.html

I am surprised to see that the GPU code is not faster.

can you try with a larger dataset ?

>

**Comment by GeoffNN** [comment]
Thanks! I just saw, and was looking at your dataset scripts. I'm not sure RCV1 / 20newsgroups are the best to showcase GPU acceleration since they're sparse. I'm curious though, so I'll try those first.

**Comment by GeoffNN** [comment]
Actually my code currently doesn't work for sparse torch.Tensors.

Would you have a suggestion for a good non sparse benchmark dataset for L2 logreg ?

**Comment by agramfort** [comment]
maybe covertype ?

see https://scikit-learn.org/stable/auto_examples/kernel_approximation/plot_scalable_poly_kernels.html#sphx-glr-auto-examples-kernel-approximation-plot-scalable-poly-kernels-py

**Comment by GeoffNN** [comment]
I submitted a results PR with CovType. The results I get are a bit strange...

**Comment by agramfort** [comment]
Where did you add the covertype dataset ? Shall we push it here?

**Comment by GeoffNN** [comment]
Pushed, with the madelon dataset as well (a lot of duplicate code between the two)

File: datasets/covtype.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+with safe_import_context() as import_ctx:
+    import numpy as np
+    from copt.datasets import load_covtype
+
+
+class Dataset(BaseDataset):
+    name = "covtype"
+
+    install_cmd = 'conda'
+    requirements = ['pip:copt']
```
Comment (agramfort) [inline datasets/covtype.py:13]: I would have used the sklearn fetcher to avoid having to maintaining 2 fetchers

File: datasets/covtype.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    def __init__(self):
+        self.X, self.y = None, None
+
+    def get_data(self):
+
+        if self.X is None:
+            self.X, self.y = load_covtype()
+            self.X = np.array(self.X.todense())
```
Comment (agramfort) [inline datasets/covtype.py:22]: covtype is sparse otherwise????

File: datasets/madelon.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class Dataset(BaseDataset):
+    name = "madelon"
+
+    install_cmd = 'conda'
+    requirements = ['pip:copt']
+
+    def __init__(self):
+        self.X, self.y = None, None
```
Comment (agramfort) [inline datasets/madelon.py:16]: maybe the init iis not necessary here?

File: datasets/madelon.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    def get_data(self):
+
+        if self.X is None:
```
Comment (agramfort) [inline datasets/madelon.py:20]: is this if necessary?

File: datasets/covtype.py (hunk 6 above)
Comment (GeoffNN) [reply datasets/covtype.py:22]: Yes, it's returned as a sparse_csr matrix

**Comment by agramfort** [comment]
are you sure?

```
In [9]:
   ...: from sklearn.datasets import fetch_covtype

In [10]: X, y = fetch_covtype(return_X_y=True)

In [11]: type(X)
Out[11]: numpy.ndarray
```

**Comment by GeoffNN** [comment]
Ah --  copt converts it to sparse then. I'll change it back to sklearn's dataloader. Thanks!

File: datasets/madelon.py (hunk 8 above)
Comment (GeoffNN) [reply datasets/madelon.py:20]: Removed

File: datasets/madelon.py (hunk 7 above)
Comment (GeoffNN) [reply datasets/madelon.py:16]: Removed

File: datasets/covtype.py (hunk 5 above)
Comment (GeoffNN) [reply datasets/covtype.py:12]: I'll fix the madelon dataset to do the same with make_classification

Comment (agramfort) [reply datasets/covtype.py:12]: if you fetch with sklearn then the dep is wrong

File: datasets/madelon.py (hunk 7 above)
Comment (agramfort) [reply datasets/madelon.py:16]: not yet

File: datasets/madelon.py (hunk 8 above)
Comment (agramfort) [reply datasets/madelon.py:20]: not yet

File: datasets/madelon.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+with safe_import_context() as import_ctx:
+    import numpy as np
+    from copt.datasets import load_madelon
+
+
+class Dataset(BaseDataset):
+    name = "madelon"
+
+    install_cmd = 'conda'
+    requirements = ['pip:copt']
```
Comment (agramfort) [inline datasets/madelon.py:13]: it should be requiring scikit-learn not copt

**Comment by tomMoral** [comment]
The CI should be fixed now. I restarted the test to see if the error disapear.

**Comment by GeoffNN** [comment]
@tomMoral I think they're all passing except the lightning install, which hasn't been touched by this PR

**Comment by agramfort** [comment]
results don't make much sense in https://benchopt.github.io/results/benchmark_logreg_l2_benchopt_run_2021-03-16_15h13m26.html

it's because you needed to have y = 1 or -1 given how the objective is computed.

once you acknowledge this can you send a PR to results repo to remove the spurious benchmarks I merged?

**Comment by agramfort** [comment]
thx @GeoffNN !

can you update the result benchmarks  and submit a PR to remove the old files?

thx again !

**Comment by GeoffNN** [comment]
Good catch, thank you!

**Comment by GeoffNN** [comment]
Updated the results and removed the former files! Results make a lot more sense now.

**Comment by agramfort** [comment]
Can you share the images you get ? https://benchopt.github.io/results/benchmark_logreg_l2_benchopt_run_2021-03-18_02h11m46.html Is weird
